grove/simon/simon.py

```diff
@@ -0,0 +1,450 @@
+            # set z to the bitwise xor of z and the current row
+            # as it will be guaranteed to still be orthogonal to s
+            if msb_row == msb_z:
+                z = np.array([z[i] ^ row[i] for i in xrange(n)])
+                got_to_end = False
+                break
+            # if the row has a greater msb than z,
+            # then this is the row to z insert above
+            elif msb_row > msb_z:
+                W = np.insert(W, row_num, z, 0)
+                got_to_end = False
+                break
+        # if z has a greater msb than all rows,
+        # insert it to the bottom of the array
+        if got_to_end:
+            W = np.insert(W, len(W), z, 0)
+
+    return W
+
+
+def make_square_row_echelon(W):
+    """
+    Make :math:`\\mathbf{\\mathit{W}}`
+    into a square matrix for Simon's algorithm, satisfying a few criteria.
+
+    :param 2darray W: an :math:`(n-1)\\times n` array of 0s and 1s in
+                      row-echelon form such that all rows are orthogonal
+                      to some length :math:`n` vector
+                      of 0s and 1s :math:`\\mathbf{s}`.
+    :return: a two-element tuple. The first element is
+             an :math:`n\\times n` square array identical
+             to :math:`\\mathbf{\\mathit{W}}` except with one row added.
+             That row is chosen to keep :math:`\\mathbf{\\mathit{W}}`
+             in row-echelon form. The second element is the row that
+             the new row is in, where the top row is at index 0.
+    :rtype: tuple
+    """
+    n = len(W) + 1
+
+    # Generate one final vector that is not orthonormal to the mask s
+    # can do by adding a vector with a single 1
+    # that can be inserted so that diag(W) is all ones
+    insert_row_num = 0
+    while insert_row_num < n - 1 and W[insert_row_num][insert_row_num] == 1:
+        insert_row_num += 1
+
+    new_row = np.zeros(shape=(n,))
+    new_row[insert_row_num] = 1
+    W = np.insert(W, insert_row_num, new_row, 0)
+
+    return W, insert_row_num
+
+
+def binary_back_substitute(W, s):
+    """
+    Perform back substitution on a binary system of equations.
+
+    Finds the :math:`\\mathbf{x}` such that
+    :math:`\\mathbf{\\mathit{W}}\\mathbf{x}=\\mathbf{s}`,
+    where all arithmetic is taken bitwise and modulo 2.
+
+    :param 2darray W: A square :math:`n\\times n` matrix of 0s and 1s,
+              in row-echelon form
+    :param list(int) s: An :math:`n\\times 1` vector of 0s and 1s
+    :return: The :math:`n\\times 1` vector of 0s and 1s that solves the above
+             system of equations.
+    :rtype: list(int)
+    """
+    # iterate backwards, starting from second to last row for back-substitution
+    s_copy = np.array(s)
+    n = len(s)
+    for row_num in xrange(n - 2, -1, -1):
+        row = W[row_num]
+        for col_num in xrange(row_num + 1, n):
+            if row[col_num] == 1:
+                s_copy[row_num] = int(s_copy[row_num]) ^ int(s_copy[col_num])
```

Comment:
I myself would prefer this to be `(a + b) % 2`, even if xor is binary addition mod 2.